src/arch_cuda.jl

```diff
@@ -0,0 +1,108 @@
+AC.on_architecture(::CUDAGPU, a::CuArray) = a
+AC.on_architecture(::CUDAGPU, a::BitArray) = CuArray(a)
+AC.on_architecture(::CUDAGPU, a::SubArray{<:Any, <:Any, <:CuArray}) = a
+AC.on_architecture(::CUDAGPU, a::SubArray{<:Any, <:Any, <:Array}) = CuArray(a)
+AC.on_architecture(::AC.CPU, a::SubArray{<:Any, <:Any, <:CuArray}) = Array(a)
+AC.on_architecture(::CUDAGPU, a::StepRangeLen) = a
+
+## GPU to GPU copy of contiguous data
+@inline function AC.device_copy_to!(dst::CuArray, src::CuArray; async::Bool = false)
+    n = length(src)
+    context!(context(src)) do
+        GC.@preserve src dst begin
+            unsafe_copyto!(pointer(dst, 1), pointer(src, 1), n; async)
+        end
+    end
+    return dst
+end
+
+@inline Architectures.unsafe_free!(a::CuArray) = CUDA.unsafe_free!(a)
+
+@inline AC.constructors(::AC.GPU{CUDABackend}, A::SparseMatrixCSC) = (CuArray(A.colptr), CuArray(A.rowval), CuArray(A.nzval),  (A.m, A.n))
+@inline AC.constructors(::AC.CPU, A::CuSparseMatrixCSC) = (A.dims[1], A.dims[2], Int64.(Array(A.colPtr)), Int64.(Array(A.rowVal)), Array(A.nzVal))
+@inline AC.constructors(::AC.GPU{CUDABackend}, A::CuSparseMatrixCSC) = (A.colPtr, A.rowVal, A.nzVal,  A.dims)
+
+@inline AC.unpack_constructors(::AC.CPU, constr::Tuple) = (constr[3], constr[4], constr[5])
+@inline AC.copy_unpack_constructors(::AC.CPU, constr::Tuple) = deepcopy((constr[3], constr[4], constr[5]))
+
+@inline AC.arch_sparse_matrix(::AC.GPU{CUDABackend}, constr::Tuple) = CuSparseMatrixCSC(constr...)
+@inline AC.arch_sparse_matrix(::AC.CPU, A::CuSparseMatrixCSC)   = SparseMatrixCSC(constructors(AC.CPU(), A)...)
+@inline AC.arch_sparse_matrix(::AC.GPU{CUDABackend}, A::SparseMatrixCSC)     = CuSparseMatrixCSC(constructors(AC.GPU(), A)...)
+
+@inline AC.arch_sparse_matrix(::AC.GPU{CUDABackend}, A::CuSparseMatrixCSC) = A
+
+@inline AC.convert_to_device(::CUDAGPU, args) = CUDA.cudaconvert(args)
+@inline AC.convert_to_device(::CUDAGPU, args::Tuple) = map(CUDA.cudaconvert, args)
+
+
+BC.validate_boundary_condition_architecture(::CuArray, ::AC.GPU, bc, side) = nothing
+
+BC.validate_boundary_condition_architecture(::CuArray, ::AC.CPU, bc, side) =
+    throw(ArgumentError("$side $bc must use `Array` rather than `CuArray` on CPU architectures!"))
+
+function SO.plan_forward_transform(A::CuArray, ::Union{BC.Bounded, BC.Periodic}, dims, planner_flag)
+    length(dims) == 0 && return nothing
+    return CUDA.CUFFT.plan_fft!(A, dims)
+end
+
+function SO.plan_backward_transform(A::CuArray, ::Union{BC.Bounded, BC.Periodic}, dims, planner_flag)
+    length(dims) == 0 && return nothing
+    return CUDA.CUFFT.plan_ifft!(A, dims)
+end
+
+# CUDA version, the indices are passed implicitly
+CUDA.@device_override @inline function __validindex(ctx::UT.MappedCompilerMetadata)
+    if __dynamic_checkbounds(ctx)
+        index = @inbounds UT.linear_expand(__iterspace(ctx), blockIdx().x, threadIdx().x)
+        return index ≤ UT.__linear_ndrange(ctx)
+    else
+        return true
+    end
+end
```

Comment:
Uhm. what the heck is this thing?

Reply:
because KA doesn't support index maps, right @simone-silvestri ?

Reply:
Yeah, it was this PR #3920. I am happy to try to implement it in KA if you want @vchuravy, I thought the conclusion was you didn't want it there.

Reply:
Ah sorry, I didn't want to stifle development in that direction. If I remember my thoughts from back then my feeling was more that I was the wrong person to implement this since I didn't have an active need.

But since then multiple people have build something similar, so having a sensible solution in KA would make me happy